File: C++/VideoFilter.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <string>
#include <cmath>

using namespace std;
using namespace cv;

string printTime(double seconds);

int main() {

    time_t timer1, timer2;
    time(&timer1);

    cout << "OpenCV Version: " << CV_VERSION << endl;
    VideoCapture capture("test_drive.mp4");

    if (!capture.isOpened()) {
        cout << "Could not open file" << endl;
        return -1;
    }
    int width = capture.get(3);
    int height = capture.get(4);
    double fps = capture.get(5);
    int frame_count = capture.get(7);
    cout << "Frame Count:" << frame_count << endl;

    int totalNumPixels = width * height;

    double duration = frame_count / fps;
    float threshold = 0.4;

    cout << "*******************************************" << endl;
    printf( "Video Dimensions: %d x %d \n", width, height );
    printf( "Video Framerate: %f \n", fps );
    printf( "Video Duration: %s \n", printTime(duration).c_str());
    cout << "*******************************************" << endl;

    int codec = VideoWriter::fourcc('m', 'p', '4', 'v');
    cout << codec << endl;
    VideoWriter output("output.mp4" , codec, fps, Size(width, height));

    if (!output.isOpened()) {
        cout << "Could not open output file" << endl;
        return -1;
    }

    int cutCount = 0;
    int frameCount = 0;
    int failedFrames = 0;
    bool lastRet = false;
    int counter = 0;

    while (capture.isOpened()) {
        count++;

        Mat frame, frame_HSV, frame_threshold;
        bool isSuccess = capture.read(frame);
        capture >> frame;
        if (frame.empty()) {
            cout << "frame empty" << endl;
            break;
        }

        if (isSuccess) {
            frameCount++;

            cvtColor( frame, frame_HSV, COLOR_BGR2HSV);
            inRange(frame_HSV, Scalar(100, 50, 50), Scalar(130, 255, 255), frame_threshold);

            imshow("Original", frame);
            imshow("Masked", frame_threshold);

            double numWhitePixels = countNonZero(frame_threshold);
            double numBlackPixels = totalNumPixels - numWhitePixels;

            cout << "Num Black Pixels: " << numBlackPixels << endl;
            cout << "Num White Pixels: " << numWhitePixels << endl;
            cout << endl;

            cout << numWhitePixels / totalNumPixels << endl;
            if ((numWhitePixels / totalNumPixels) > threshold) {
                cout << "Frame " << frameCount << " Cut" << endl;

                cutCount++;
                cout << cutCount << endl;
                lastRet = true;
            }
            else {
                cout << "Frame " << frameCount << " Retained" << endl;

                output.write(frame);

                lastRet = false;
            }

        }
        if (!isSuccess) {
            cout << "Video has ended" << endl;
            break;
        }

        int key = waitKey(20);
        if (key == 'q') {
            cout << "Process ended by user" << endl;
            break;
        }

    }

    output.release();
    capture.release();
    destroyAllWindows();

    cout << "Video Cutting Complete" << endl;
    printf("%d of %d total frames (%05.2f%%) were cut \n", cutCount, frameCount, (static_cast<float>(cutCount)/frameCount)*100.0);

    time(&timer2);
    float elapsedTime = difftime(timer2, timer1);

    cout << "ELAPSED: " << printTime(elapsedTime) << endl;
    cout << "count: " << counter;

    return 0;

}

string printTime(double seconds) {
    return format("%02d:%02d:%05.2f", (static_cast<int>(seconds / 3600)), (static_cast<int>(seconds / 60)), fmod(seconds , 60.0) );
}

File: python/VideoFilter.py
import cv2
import numpy as np
import sys
import time
import itertools
import threading

start = time.perf_counter()

# Retain: frames or clips that were 'kept in' from the original video, i.e. added to the output file
# Note that audio is not retained in output video file, if audio is desired it will need to be readded manually
# Audio could be lined up by using the timestamps of clips used from the original video, printed at the end of the program

fileName = sys.argv[1]          # when running script first argument is the filename and the second is the output filename and third is the blue threshold
outputFileName = sys.argv[2]    # output filename
threshold = float(sys.argv[3])  # what fraction of frame that must be blue in order to be cut (ex: more than 40% -> 0.4, more than 90% -> 0.9)

# Because cutting videos may take a while, the loading animation tells the user that the process has not frozen
done = False
def animate():                  # Loading animation that displays while video is being cut
    for c in itertools.cycle(['|', '/', '-', '\\']):
        if done:
            break
        sys.stdout.write('\rCutting Video ' + c)
        sys.stdout.flush()
        time.sleep(0.1)

def printTime(seconds):
    return f"{int(seconds//3600):02d}:{int(seconds//60):02d}:{round(seconds % 60, 2):05.2f}"

capture = cv2.VideoCapture(fileName)
if (capture.isOpened() == False):
    print("Could not open file", fileName)

width = int(capture.get(3))     # gets information about video using the cv2 'get' function
height = int(capture.get(4))
fps = capture.get(5)
frameCount = int(capture.get(7))
TotalNumPixels = height * width

fourcc = cv2.VideoWriter_fourcc(*'mp4v')    # If output is desired in .avi format, change *'mp4v' to *'xvid' (output filename must also be .avi)
out = cv2.VideoWriter(outputFileName, fourcc, fps, (width,height))
duration = printTime(frameCount/fps)

print("\n******************** Video Info ********************\n")
print(f"Video Dimensions: {width} x {height}")
print(f"Video Framerate: {fps}")
print(f"Video Duration: {duration}")
print("\n****************************************************\n")

t = threading.Thread(target=animate) # Begins loading animation
t.start()

frameCount = 0
cutCount = 0
lastRet = False
timeStamps = []

while True:
    ret, frame = capture.read()

    if ret == False:            # ret returns false when there are no frames left in the input video
        break                   # exits program

    hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)

    lowerBlue = np.array([100, 50, 50])
    upperBlue = np.array([130, 255, 255])

    mask = cv2.inRange(hsv, lowerBlue, upperBlue)
    numWhitePixels = np.sum(mask == 255)
    numBlackPixels = np.sum(mask == 0)

    # cv2.imshow('frame', frame) # will display original video in real time
    # cv2.imshow('mask', frame)   # will display masked video in real time - Note: these slow down program

    if ((numWhitePixels / TotalNumPixels ) > threshold):    # Cuts frame if more than [threshold] of frame is blue
        #print(f"Frame {frameCount}: REMOVED")
        cutCount += 1
        if (lastRet == True):                               # If last clip was not cut out and this one is, record that frame
            timeStamps.append(frameCount)
        lastRet = False
    else:
        #print(f"Frame {frameCount}: RETAINED")
        out.write(frame)
        if (lastRet == False):                              # If last clip was cut out and this clip is not, record that frame
            timeStamps.append(frameCount)                   # These record where the 'clips' retained start and end, used to find the timestamps
        lastRet = True                                              # at the end so that the clips used can be found in the original video

    frameCount += 1

if (lastRet == True):
    timeStamps.append(frameCount)

end = time.perf_counter()
done = True

print("\nVideo Cutting Complete")
print("\n******************** Edit Info *********************\n")
print(f"Blue Frame Threshold: {threshold*100}%")
print(f"{cutCount} of {frameCount} total frames ({round((cutCount/frameCount)*100,2)}%) were cut")
print(f"Time Elapsed: {printTime(end-start)}")
print("\n****************************************************\n")

capture.release()
out.release()
cv2.destroyAllWindows()

print("\n******************** Timestamps *********************\n")
print("Timestamps of Clips Retained from Original Video:")
for i, stamp in enumerate(timeStamps):
    if ((i % 2) == 0):
        print()
    elif ((i % 2) == 1):
        print(" - ", end='')
    print(printTime(stamp/fps), end='')
print("\n")
